plugins/modules/foreman_hostgroup.py

```diff
@@ -408,12 +414,34 @@ def main():
 
     parameters = entity_dict.get('parameters')
 
+    puppetclasses = entity_dict.pop('puppetclasses', None)
+    current_puppetclasses = entity.pop('puppetclasses', None) if entity else []
+
     hostgroup = module.ensure_entity('hostgroups', entity_dict, entity)
 
     if hostgroup:
         scope = {'hostgroup_id': hostgroup['id']}
         module.ensure_scoped_parameters(scope, entity, parameters)
 
+    if not module.desired_absent and 'environment_id' in hostgroup:
+        if puppetclasses is not None:
+            current_puppetclasses = [p['id'] for p in current_puppetclasses]
+            for puppet_class_name in puppetclasses:
+                scope = {'environment_id': hostgroup['environment_id']}
+                results = module.list_resource("puppetclasses", params=scope, search='name="{0}"'.format(puppet_class_name))
+                if len(results) == 1 and len(list(results.values())[0]) == 1:
+                    puppet_class_id = list(results.values())[0][0]['id']
+                    if puppet_class_id in current_puppetclasses:
+                        current_puppetclasses.remove(puppet_class_id)
+                    else:
+                        payload = {'hostgroup_id': hostgroup['id'], 'puppetclass_id': puppet_class_id}
+                        module.ensure_entity('hostgroup_classes', None, None, params=payload, state='present')
```

Comment:
Now, you just put all the information into the `payload`, which is probably ok, but you should still pass it an (now empty as in `{}`) entity_spec, because otherwise it will use the hostgroup_spec with unforeseeable results.
 The same holds true for line 443.
Btw, if you have seen this pattern elsewhere, we should review that also.